lib/Canonicalize/MakeModelGEPPass.cpp

```diff
@@ -1478,30 +1451,33 @@ computeBestInUnion(const model::QualifiedType &BaseUnion,
 }
 
 static RecursiveCoroutine<ModelGEPReplacementInfo>
-computeBest(const model::QualifiedType &BaseType,
+computeBest(const model::UpcastableType &BaseType,
             const IRSummation &IRSum,
-            const std::optional<model::QualifiedType> &AccessedTypeOnIR,
+            const model::UpcastableType &AccessedTypeOnIR,
             model::VerifyHelper &VH) {
   revng_log(ModelGEPLog, "Computing Best ModelGEP for IRSum: " << IRSum);
-  namespace TDKind = model::TypeDefinitionKind;
-  revng_assert(not BaseType.isVoid()
-               and not BaseType.is(TDKind::RawFunctionDefinition)
-               and not BaseType.is(TDKind::CABIFunctionDefinition));
+
+  auto UnwrappedBaseType = model::copyType(*BaseType->skipConstAndTypedefs());
+  revng_assert(not UnwrappedBaseType->isVoid()
+               and not UnwrappedBaseType->isPrototype());
 
   // This Result models no access, and leaves all the IRSum as a mismatch.
   // It's handful for easy bail out for obvious failures to traverse this
   // BaseType properly.
-  ModelGEPReplacementInfo Result{ BaseType, {}, IRSum, BaseType };
+  ModelGEPReplacementInfo Result(/* Base = */ BaseType.copy(),
+                                 /* Indices = */ {},
+                                 /* Mismatch = */ IRSum,
+                                 /* Accessed = */ BaseType.copy());
 
   // If we've reached a type that cannot be "traversed" by a ModelGEP we have
-  // nothing to do. This might be subject to cheange in the future when we'll
+  // nothing to do. This might be subject to change in the future when we'll
   // need to support "accessing" pointer with the square brackets as if they
   // were arrays.
-  if (BaseType.isPrimitive() or BaseType.isPointer()
-      or BaseType.is(model::TypeDefinitionKind::EnumDefinition)) {
+  if (UnwrappedBaseType->isPrimitive() || UnwrappedBaseType->isPointer()
+      || UnwrappedBaseType->isEnum()) {
```

Comment:
Can we use `isScalar`? Maybe backlog this.

Reply:
As far as I remember, enums are not considered primitives as of now. Should they be?

Reply:
Alternatively, we can have `isScalarOrEnum` or something.

Reply:
On a second thought, this is not the right time for this kind of fixes.
Maybe add a TODO like "consider using `isScalar`".

Reply:
We could in fact use `isScalar` here, my bad.